### PR TITLE
Rule information during integration testing 

### PR DESCRIPTION
**Pull request by Soph1514**
Enable getting all the rules in order of their application to a an essence problem and store them in a file. When cargo test is run, in addition to getting the parsed, rewritten model and the solutions, a list of rules for each test case will be generated and tested against the expected file.

### CODE REVIEW

**Comment by ozgurakgun** [comment]
I think this prints the name of the rule only, and not the expression it was applied to.

See https://github.com/conjure-cp/conjure-oxide/blob/c00a2b5f26b658d9bb92707b7de6f4d58abe40e3/crates/conjure_core/src/rule_engine/rewrite.rs#L303 and https://github.com/conjure-cp/conjure-oxide/blob/c00a2b5f26b658d9bb92707b7de6f4d58abe40e3/crates/conjure_core/src/rule_engine/rewrite.rs#L317 -- I think these are closer to the output we want. You will have to enable them through the log library though.

**Comment by niklasdewally** [comment]
LGTM! Just a few comments below.

* I agree that we should use Rusts logging functionality where possible.  This gives us lots of flexibility down the line over the format and chattiness of the information. If you use the default logger on trace level, this should already go both to a file and stderr.

* **Optional:** You could also create a seperate log target for the rule logger, so that you can make it show up separately to the other logs, saved in a different file, say "rule_engine" before logs so we know what subsystem it came from, etc. 

* **Optional:** Writing a simple macro, `log_rule_app!` to wrap the call to whatever underlying logging / writing you do might be good. This way, you can change how you format these in one place, and your actual logging lines should be easier to write.  See the parser logging macros below.

* **Nit:** Ideally we should log to stderr not stdout (`eprintln!` instead of `println!`). I know that we aren't yet consistent on this across the codebase though!

For example, here's some stuff we do in the parser:

https://github.com/conjure-cp/conjure-oxide/blob/c00a2b5f26b658d9bb92707b7de6f4d58abe40e3/crates/conjure_core/src/parse/parse_model.rs#L14-L24

https://github.com/conjure-cp/conjure-oxide/blob/c00a2b5f26b658d9bb92707b7de6f4d58abe40e3/crates/conjure_core/src/parse/parse_model.rs#L304-L308

https://github.com/conjure-cp/conjure-oxide/blob/c00a2b5f26b658d9bb92707b7de6f4d58abe40e3/conjure_oxide/src/main.rs#L100-L104

File: crates/conjure_core/src/rule_engine/rewrite.rs [hunk 1]
```diff
@@ -347,9 +351,13 @@ fn apply_all_rules<'a>(
 /// }
 ///
 fn choose_rewrite(results: &[RuleResult]) -> Option<Reduction> {
+    if results.len() > 1 {
+        bug!("There should be only 1 rule that can be applied to an expression at a time");
```
Comment (ozgurakgun) [inline crates/conjure_core/src/rule_engine/rewrite.rs:355]: looks good! if it actually happens we would probably also want to print the names of the rules etc.

Comment (ozgurakgun) [reply crates/conjure_core/src/rule_engine/rewrite.rs:355]: also: let's see if it code coverage correctly ignores this line.

Comment (Soph1514) [reply crates/conjure_core/src/rule_engine/rewrite.rs:355]: This caused a lot of integration tests to fail, does that mean there is a problem with our rules or the if statement?

Comment (Soph1514) [reply crates/conjure_core/src/rule_engine/rewrite.rs:355]: Of course! I will have it print to the console for now and once I figure out the logging will implement that there

Comment (ozgurakgun) [reply crates/conjure_core/src/rule_engine/rewrite.rs:355]: oh I was just meaning string concatanation to make the message you pass to bug! more descriptive. I'll check the failing tests.

Comment (ozgurakgun) [reply crates/conjure_core/src/rule_engine/rewrite.rs:355]: yep, we need that message to tell us a bit more :)

it looks like multiple rules were applicable all over the place. good if statement!

Comment (ozgurakgun) [reply crates/conjure_core/src/rule_engine/rewrite.rs:355]: I suggest we separate the two changes, as they are orthogonal. adding tracing (which is what this PR started as) is one change, adding the if statement and working on its implications should be another PR.

**Comment by Soph1514** [comment]
To enable the following line `log::trace!(target: "file", "Rule applicable: {:?}, to Expression: {:?}, resulting in: {:?}", rule, expression, red.new_expression); ` do I need to create a new builder in the main function? Or work with `generated_tests.rs` since I want the logging and file creation to happen when `cargo test `is run.

**Comment by niklasdewally** [comment]
> To enable the following line `log::trace!(target: "file", "Rule applicable: {:?}, to Expression: {:?}, resulting in: {:?}", rule, expression, red.new_expression); ` do I need to create a new builder in the main function? Or work with `generated_tests.rs` since I want the logging and file creation to happen when `cargo test `is run.

That sounds about right 👍 

If the current builders don't suit, lots of logging libraries (that support `log`) exist that do things in different ways.For example https://docs.rs/env_logger/latest/env_logger/

Note that tests run in parallel, so you might want to wrap your builder in a OnceLock: https://doc.rust-lang.org/std/sync/struct.OnceLock.html.

This provides a way to guarantee that a variable is initialised exactly once.

**Comment by niklasdewally** [comment]
Seems that lazylock exists now which seems much simpler :) 
 https://doc.rust-lang.org/std/sync/struct.LazyLock.html

**Comment by ozgurakgun** [comment]
We want a separate log file per integration test, that is checked like we checked the other 3 files (expected/output style). Keep that in mind as well when choosing a logging library.

**Comment by Soph1514** [comment]
I have managed to capture the trace from `rewrite_model `to files for all integration tests but only when the function is called inside the tracing::subscriber scope which then means we do not have access to the model itself. I couldn't create a global subscriber because it says `a global default trace dispatcher has already been set`. Where are we creating a global one because I could not find it in the code? 

`generated_tests.rs`
```
let (subscriber, _guard) = create_scoped_subscriber(path, essence_base);
    tracing::subscriber::with_default(subscriber, || {
        // Create a span for the trace
        let test_span = span!(target: "rule_engine", Level::TRACE, "test_span");
        let _enter = test_span.enter();
        trace!(target: "rule_engine", "hello");
    });

    let model = rewrite_model(&model, &rule_sets)?;
    if verbose {
        println!("Rewritten model: {:#?}", model)
    }
```

**Comment by niklasdewally** [comment]
My guess is that the above code is global not thread-local.

**Comment by Soph1514** [comment]
This does not produce an error though, only when I use tracing::subscriber.set_global it panics and says that a global one already exists. I was just wondering where is the global one defined in our repo so I could maybe try adding mother target to it?

**Comment by niklasdewally** [comment]
Sorry, I deleted my message as I looked deeper into the library you used and it does seem like the right tool for the job!

**Comment by niklasdewally** [comment]
> This does not produce an error though, only when I use tracing::subscriber.set_global it panics and says that a global one already exists. I was just wondering where is the global one defined in our repo so I could maybe try adding mother target to it?

Do we need a global logger? Does having the scoped ones as you've shown above work?

**Comment by Soph1514** [comment]
> > This does not produce an error though, only when I use tracing::subscriber.set_global it panics and says that a global one already exists. I was just wondering where is the global one defined in our repo so I could maybe try adding mother target to it?
> 
> Do we need a global logger? Does having the scoped ones as you've shown above work?

It only works if I put the function call inside the brackets of tracing::subscriber which leads to many issues, like inability to use` ?` and the `model` being dropped after the scope ends. I tried wrapping the last 2 test stages into the subscriber but then the compiler complains about asserts. I am just not sure how to approach this and I did not want to mess with the already established code much. Perpahs the `model` can be initialised outside first somehow and made mutable?

**Comment by niklasdewally** [comment]
> > > This does not produce an error though, only when I use tracing::subscriber.set_global it panics and says that a global one already exists. I was just wondering where is the global one defined in our repo so I could maybe try adding mother target to it?
> > 
> > 
> > Do we need a global logger? Does having the scoped ones as you've shown above work?
> 
> It only works if I put the function call inside the brackets of tracing::subscriber which leads to many issues, like inability to use` ?` and the `model` being dropped after the scope ends. I tried wrapping the last 2 test stages into the subscriber but then the compiler complains about asserts. I am just not sure how to approach this and I did not want to mess with the already established code much. Perpahs the `model` can be initialised outside first somehow and made mutable?


It might be easier to use the logger for the entire lifetime of the thread, instead of just for the rewrite model bit? `integration_test` wraps the test code with setup/teardown code; if you put the logger in there, it should last for the entire duration of the test. However, this would mean that we could only check the logs at the end of test, during the teardown stage; not sure if this is suitable in this case?


(I am also not too familiar with `tracing`, so there might be a technical reason why this wouldn't work)

**Comment by niklasdewally** [comment]
> Perpahs the model can be initialised outside first somehow and made mutable

With closures the usual solution is to use smart pointers /  `Arc` as so:

```rs
let model_ptr1 = Arc::new(my_model);
let model_ptr2 = model_ptr1.clone();

do_some_stuff(move ||{
  check_my_model(*model_ptr1)
})

do_some_more_stuff(*model_ptr2)
```

The above code makes 2 pointers for the model, then gives ownership of the first pointer to the inside of do_some_stuff, to do some stuff with.

The main code keeps ownership of the second pointer, allowing it to also access the model.

**Comment by niklasdewally** [comment]
For mutability, you would need to add a `RefCell` or a `Mutex` inside the `Arc` (depending on how thread safe you want to be).

This can be messy, so I will defer the giving of an example to stack overflow:(https://stackoverflow.com/questions/31373255/how-do-i-share-a-mutable-object-between-threads-using-arc):

```rs
fn main() {
    let shared_stats = Arc::new(Mutex::new(Stats));
    let threads = 5;
    for _ in 0..threads {
        let my_stats = shared_stats.clone();
        thread::spawn(move || {
            let mut shared = my_stats.lock().unwrap();
            shared.add_stats(&Stats);
        });
        // Note: Immediately joining, no multithreading happening!
        // THIS WAS A LIE, see below
    }
```

**Comment by ozgurakgun** [comment]
This is looking good so far.

Should we change the filenames to something like:

BASENAME.expected-rule-trace.txt
BASENAME.generated-rule-trace.txt

like the other 3 files?

We would then update the accept code paths as well.

**Comment by Soph1514** [comment]
That would be my next step. Just wanted to make it work first:)

**Review by niklasdewally — COMMENTED** [review]
Looks good so far, just a few comments.

Cloning this branch caused my git submodules to freakout and minion to fail the build, but this could just be me.

**Comment by Soph1514** [comment]
@niklasdewally Hey! I am the following error. I thought it was because my branch was not up to date but it happened even after I merged. 
image

**Comment by ozgurakgun** [comment]
in the github action there is a different error:

```
error: duplicate key `tracing-subscriber` in table `dependencies`
  --> conjure_oxide/Cargo.toml:41:1
   |
41 | tracing-subscriber = { version = "0.3.18", features = ["ansi", "env-filter", "json"] }
   | ^
   |
```

this might be because of a duplicate entry in Cargo.toml (can happen during a merge)

regarding the error you posted, I would make sure you are on the commit you intend to be on, then do `git clean -dfx .` to remove untracked files, then try `cargo test --workspace` again.

if this works we should add it to an appropriate place in the wiki as documentation / possible troubleshooting step.

-- aside --

incidentally I have been thinking of how we could use the wiki more effectively. Maybe we can extract a pdf of the entire contents as a "book" and we all aim to make sure that is as complete as possible. This project is a good example: https://github.com/jobisoft/wikidoc

Could be a project for someone next semester...

**Comment by ozgurakgun** [comment]
another duplicate?

```
error: duplicate key `tracing` in table `dependencies`
  --> crates/conjure_core/Cargo.toml:35:1
   |
35 | tracing = "0.1.40"
   | ^
   |
error: failed to load manifest for workspace member `/Users/runner/work/conjure-oxide/conjure-oxide/conjure_oxide`
```

can you see the github action logs @Soph1514 ?

File: conjure_oxide/src/main.rs [hunk 2]
```diff
@@ -1,35 +1,36 @@
+// (niklasdewally): temporary, gut this if you want!
+
```
Comment (niklasdewally) [inline conjure_oxide/src/main.rs:2]:
```suggestion

```

File: conjure_oxide/src/main.rs
Comment (niklasdewally) [inline conjure_oxide/src/main.rs:1]: Did you change `main.rs` at all?  These changes just seem to undo changes to `main.rs` made in other commits.

File: readme 2.md
Comment (niklasdewally) [inline readme 2.md:1]: Delete this

File: crates/conjure_core/src/rule_engine/rewrite.rs [hunk 3]
```diff
@@ -431,3 +436,10 @@ fn check_priority<'a>(
         \n Rules{:?}", initial_expr, new_expr, rules)
     }
 }
+fn trace_rules(results: &[RuleResult], expression: Expression) {
+    if !results.is_empty() {
+        let rule = results[0].rule;
+        let new_expression = results[0].reduction.new_expression.clone();
+        trace!(target: "rule_engine", "Rule applied: {:?}, to Expression: {:?}, resulting in: {:?}", rule.name, expression, new_expression);
```
Comment (niklasdewally) [inline crates/conjure_core/src/rule_engine/rewrite.rs:443]:
```suggestion
        trace!(target: "rule_engine", "Rule applied: {:?}, to Expression: {}, resulting in: {}", rule.name, expression, new_expression);
```

Expression implements Display, so we don't need to debug print here.
I would take a look at what we do on main here.

File: conjure_oxide/src/main.rs
Comment (niklasdewally) [reply conjure_oxide/src/main.rs:1]:
```
git checkout main -- conjure_oxide/src/main.rs
```

**Comment by niklasdewally** [comment]
@Soph1514 Once you've addressed a code review comment, could you click the `resolve conversation` button? It would be helpful to see what's done/still needs to be done.

**Review by niklasdewally — APPROVED** [review]
Reposting my Teams message here for visibility:

LGTM, just needs some git cleanup (remove changes to main.rs and readme2.txt).  After that, I would mark this ready to review and try to get this merged.

It would be better now that it works to merge this and improve things one by one in seperate PRs.

**Comment by Soph1514** [comment]
I am quite confused why the build test fails. When I run `cargo test --workshop` on my updated branch it passes

**Review by ozgurakgun — APPROVED** [review]
Well done @Soph1514 and @a-nosnitram !

I would like to talk about #451 at some point, maybe at tomorrow's meeting, but this looks good.

**Comment by ozgurakgun** [comment]
Oops I should have squash-merged. Next time...

**Pull request merged**